### PR TITLE
[WIP] snmg python wrapper

### PR DESCRIPTION
**Pull request by Iroy30**
Added python wrapper for multi gpu pagerank algorithm.
Added mg_pagerank test function to verify end to end working

### CODE REVIEW

File: python/cugraph/mg_pagerank/mg_pagerank_wrapper.pyx [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+                     <gdf_column*> pr_ptr,
+                     <float> 0.85,#damping_factor,
+                     <int> 20 #max_iter
+                     )
+
+    data = rmm.device_array_from_ptr(<uintptr_t> pr_ptr.data,
+                                            nelem=pr_ptr.size,
+                                            dtype=np.int32)
+    x = cudf.DataFrame()
+    x['pagerank'] = cudf.Series(data)
```
Comment (afender) [inline python/cugraph/mg_pagerank/mg_pagerank_wrapper.pyx]: Let's add a place holder for vertex_indexes

File: cpp/src/structure/cugraph.cu [hunk 2]
```diff
@@ -311,3 +317,50 @@ gdf_error gdf_delete_transposed_adj_list(gdf_graph *graph) {
...
+  for(int i=0;i<N;i++)
+  {
+      total_length = total_length + src_ptrs[i].size;
+  }
+
+  int* h_result = (int*)malloc(total_length*sizeof(int));
+  int *final_result = h_result;
+  int *d_result;
+  cudaMalloc(&d_result, total_length*sizeof(int));
+
```
[the remark this answers is not in the document]
Comment (Iroy30) [reply cpp/src/structure/cugraph.cu]: Added it. Just need to change the setdevice part as discussed (where device_to_be_set = thread_id)

File: cpp/src/structure/cugraph.cu [hunk 3]
```diff
@@ -311,3 +317,50 @@ gdf_error gdf_delete_transposed_adj_list(gdf_graph *graph) {
...
+  for(int i =0;i<N;i++)
+  {
+      cudaPointerAttributes attr;
+      cudaPointerGetAttributes (&attr, src_ptrs[i].data);
+      cudaDeviceSynchronize();
+      int dev = attr.device;
+      cudaSetDevice(dev);
+      int *ans;
+      cudaMalloc(&ans, src_ptrs[i].size*sizeof(int));
```
[the remark this answers is not in the document]
Comment (Iroy30) [reply cpp/src/structure/cugraph.cu]: at the end the result is put in a gdf_column pr which is the output

File: python/cugraph/mg_pagerank/mg_pagerank_wrapper.pyx [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+import cudf
+import numpy as np
+
+def mg_pagerank(src_ptrs_info, dest_ptrs_info):
+    print("mg_pagerank.py: ", src_ptrs_info, dest_ptrs_info)
+    
+    cdef gdf_column* src_column_ptr = <gdf_column*>malloc(len(src_ptrs_info) * sizeof(gdf_column))
+    cdef gdf_column* dest_column_ptr = <gdf_column*>malloc(len(dest_ptrs_info) * sizeof(gdf_column))
+    
+    N = len(src_ptrs_info);
```
[the remark this answers is not in the document]
Comment (Iroy30) [reply python/cugraph/mg_pagerank/mg_pagerank_wrapper.pyx]: Changed

File: cpp/src/structure/cugraph.cu [hunk 5]
```diff
@@ -311,3 +317,50 @@ gdf_error gdf_delete_transposed_adj_list(gdf_graph *graph) {
...
+      dummy_Kernel<<<1,1>>>((int*)src_ptrs[i].data,(int*)dest_ptrs[i].data, src_ptrs[i].size, (int*)ans);
```
[the remark this answers is not in the document]
Comment (Iroy30) [reply cpp/src/structure/cugraph.cu]: edited

File: cpp/src/structure/cugraph.cu [hunk 6]
```diff
@@ -311,3 +317,50 @@ gdf_error gdf_delete_transposed_adj_list(gdf_graph *graph) {
   graph->transposedAdjList = nullptr;
   return GDF_SUCCESS;
 }
+
+__global__ void dummy_Kernel(int* src, int* dst, size_t n, int* res) {
+        for(int i=0;i<n;i++)
+                res[i]= src[i] + dst[i];
+}
+
+gdf_error gdf_dummy_multi_pagerank(const size_t N, gdf_column *src_ptrs, gdf_column *dest_ptrs, gdf_column *pr, const float damping_factor, const int max_iter){
```
[the remark this answers is not in the document]
Comment (Iroy30) [reply cpp/src/structure/cugraph.cu]: moved

File: cpp/src/structure/cugraph.cu [hunk 7]
```diff
@@ -311,3 +317,50 @@ gdf_error gdf_delete_transposed_adj_list(gdf_graph *graph) {
...
+      cudaPointerGetAttributes (&attr, src_ptrs[i].data);
+      cudaDeviceSynchronize();
+      int dev = attr.device;
+      cudaSetDevice(dev);
+      int *ans;
+      cudaMalloc(&ans, src_ptrs[i].size*sizeof(int));
+      dummy_Kernel<<<1,1>>>((int*)src_ptrs[i].data,(int*)dest_ptrs[i].data, src_ptrs[i].size, (int*)ans);
+      cudaDeviceSynchronize();
+      cudaMemcpy(final_result, ans, src_ptrs[i].size*sizeof(int), cudaMemcpyDeviceToHost);
+      final_result = final_result + src_ptrs[i].size;
```
[the remark this answers is not in the document]
Comment (Iroy30) [reply cpp/src/structure/cugraph.cu]: currently, the results reside in the gpu associated with the main worker. Similar to pagerank results, the output is sent back from this gpu and distributed into n partitions as dask dataframe

File: python/cugraph/graph/c_graph.pyx [hunk 8]
```diff
@@ -58,6 +58,33 @@ cdef gdf_column get_gdf_column_view(col):
 
     return c_col
 
+cdef gdf_column get_gdf_column_ptr(ipc_data_ptr, col_len):
```
[the remark this answers is not in the document]
Comment (Iroy30) [reply python/cugraph/graph/c_graph.pyx]: this function could be used for other algorithms as well apart from snmg to create a gdf_column from a given pointer

[the remark this answers is not in the document]
Comment (afender) [reply python/cugraph/graph/c_graph.pyx]: Should it go in utils.pyx then?

File: python/cugraph/mg_pagerank/mg_pagerank_wrapper.pyx (hunk 1 above)
Comment (Iroy30) [reply python/cugraph/mg_pagerank/mg_pagerank_wrapper.pyx]: added

**Pull request merged**